pkgs/c_compiler/lib/src/cbuilder/cbuilder.dart

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:logging/logging.dart';
+import 'package:native_assets_cli/native_assets_cli.dart';
+
+import 'run_cbuilder.dart';
+
+abstract class Builder {
+  Future<void> run({
+    required BuildConfig buildConfig,
+    required BuildOutput buildOutput,
+    Logger? logger,
+  });
+}
+
+/// Specification for building an artifact with a C compiler.
+class CBuilder implements Builder {
+  /// What kind of artifact to build.
+  final _CBuilderType _type;
+
+  /// Name of the library or executable to build.
+  ///
+  /// The filename will be decided by [BuildConfig.target] and
+  /// [OS.libraryFileName] or [OS.executableFileName].
+  ///
+  /// File will be placed in [BuildConfig.outDir].
+  final String name;
+
+  /// Asset identifier.
+  ///
+  /// Used to output the [BuildOutput.assets].
+  ///
+  /// If omitted, no asset will be added to the build output.
+  final String? assetName;
+
+  /// Sources to build the library or executable.
+  ///
+  /// Resolved against [BuildConfig.packageRoot].
+  ///
+  /// Used to output the [BuildOutput.dependencies].
+  final List<String> sources;
```

Comment:
Doesn't have to be in this PR, but in future you'll need to add a way of specifying header files. They're not sources because they shouldn't be passed to the clang compile command, but they do need to be included in `buildOutput.dependencies` so that we rebuild if they're modified.